codegen/core/src/main/scala/com/lightbend/akkasls/codegen/SourceGeneratorUtils.scala

```diff
@@ -181,38 +180,61 @@ object SourceGeneratorUtils {
     else dotsToCamelCase(s.tail, resultSoFar + s.head, false)
   }
 
-  class CodeBlock(val code: Seq[Any]) {
+  sealed trait CodeElement
+  object CodeElement {
+    case class StringElement(string: String) extends CodeElement
+    case class FQNElement(fqn: FullyQualifiedName) extends CodeElement
+
+    implicit def stringElement(string: String): CodeElement = StringElement(string)
+    implicit def fqnElement(fqn: FullyQualifiedName): CodeElement = FQNElement(fqn)
+    implicit def block(els: Iterable[CodeElement]): CodeElement =
+      CodeBlock(
+        // add line terminator between each two elements
+        els.toVector.flatMap(Seq(_, stringElement("\n"))).dropRight(1))
+  }
+  case class CodeBlock(code: Seq[CodeElement]) extends CodeElement {
+    import CodeElement._
+
     code.foreach(validateType)
 
     private def validateType(seq: Any): Unit = seq match {
-      case _: String             => ()
-      case _: FullyQualifiedName => ()
-      case block: CodeBlock      => block.code.foreach(validateType)
-      case s: Seq[_]             => s.foreach(validateType)
-      case other =>
-        throw new IllegalArgumentException(s"Unexpected value of type [${other.getClass}] in block: [$other]")
+      case _: StringElement => ()
```

Comment:
I think `validateType` can be removed completely now that we convert to `CodeElement` earlier